docs/src/dev/principles/cpu-abstraction.md

```diff
@@ -0,0 +1,238 @@
+# CPU Abstractions
+
+Core modules in Patina often require divergent implementations for CPU architecture specific mechanisms. This may
+be because the implementation requires assembly which cannot be compiled against other architectures, the implementation
+is only applicable to a given architecture, or both. In these cases, developers are encouraged to ensure that as much
+common code exists as possible. At the point of architecture abstraction, [trait abstraction principles](abstractions.md)
+should be used to define the common contract with the architecture specific logic. This same abstraction can then
+be used for testing of the common logic above/below the abstraction. While there is no one-size-fits-all structure to
+how this should be done, this page describes some simple and common models that should be applied to handling such
+abstractions.
+
+## Code Structure
+
+Implementations should avoid inlining architecture specific code into common logic as this creates code that is difficult
+to read, difficult to test, and is prone to inconsistent contracts between architectures. Architecture specific logic
+should be contained to explicit modules, such as `x64` or `aarch64`. The Patina crate provides macros that can be used
+to add conditional compilation in a consistent way. If the logic is not compilable on other architectures, the module
+should be conditionally compiled for that architecture.
+
+```rust
+#[cfg(target_arch = "x86_64")]
+mod x64 {
+    // A x64 specific module, intended only for platform builds.
+}
+
+#[cfg(target_arch = "aarch64")]
+mod aarch64 {
+    // A x64 specific module, intended only for platform builds.
+}
+```
+
+It should be noted that using just the architecture does cause some inconsistencies around compilation. In particular
+this will allow test compilation, but these tests may not actually be compiled/run since this requires that the
+architecture matches the host system.
+
+### Testing & Documenting

```

Comment:
Not sure if it's worth mentioning, but tests themselves can also be conditional.

```rust
#[test]
#[cfg(not(target_arch = "aarch64"))]
fn my_test() {
  // I can use arch specific things in this test, and don't have to worry about compilation issues from other arch conditional statements.
}
```